### PR TITLE
Introduces style-guide.md

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

This PR introduces style-guide.md a file where we can document our internal style guide with patterns/preferences that we want to standardize across the repo. This will not act as a resource to retro-fit our entire repo, but instead will help keep things clean/consistent moving forwards. This is a first-stab and by no means complete. Off the top of my head things missing are: unit tests/testing guidance, more specific direction on how we name variables (specifically CSS variables), how we define components (arrow functions or named/regular functions), and how we export things from files. Please feel free to add guidance around this!! For now, the workflow should just be a PR (as we normally do, and for bigger changes we can bring topics to Team Code Review on Thursdays). 

Hopefully this is helpful!

### CODE REVIEW

**Review by DesignerDave — COMMENTED** [review]
Nit: Usually the convention for read-only markdown files like this is to name them in all-caps. `STYLEGUIDE.md` or similar.

File: STYLEGUIDE.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+#### Prefer
+
+```typescript
+const handleClick = (e) => {
+    e.preventDefault()
+    setState(curr => !curr)
+}
+
+<button onClick={handleClick}>
```
Comment (spark33) [inline STYLEGUIDE.md:213]: Might also be nice to mention matching the event handler names using `on` to `handle`

**Review by spark33 — APPROVED** [review]

File: STYLEGUIDE.md (hunk 1 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:213]: Good feedback, added!

File: STYLEGUIDE.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+- [To Contribute](#To-Contribute)
+- [Typing](#Typing)
+- [Functions](#Functions)
+- [Variable naming](#Variable-naming)
+- [Variable placement](#Variable-placement)
+- [Whitespace and line breaks](#Whitespace-and-line-breaks)
+- [JSX/React](#JSX-React)
+- [CSS-in-JS](#CSS-in-JS)
+- [File Structure](https://github.com/mongodb/leafygreen-ui/blob/main/stories/Folder-Structure.stories.mdx)
+- [References](#References)
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:14]: I wonder if this would make more sense as a GH Wiki? Or maybe a separate file for each section? This feels a little long & hard to read (a bit TL;DR)

File: STYLEGUIDE.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+Ensures that TypeScript doesn't prevent people from using the values directly rather than consuming the enum.
+
+#### Prefer
+
+```typescript
+export const someConstant = {
+  optionOne = 'option 1',
+  optionTwo = 'option 2',
+} as const;
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:52]: Add how to use this in TS
```ts
export type someConstant = typeof someContent[keyof typeof someContent]
```

File: STYLEGUIDE.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+```typescript
+enum someConstant = {
+ optionOne: 'option 1',
+ optionTwo: 'option 2',
+}
+```
+
+---
+
+### When creating TS entities, place them at the top of the file, above JS
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:66]: Why?

File: STYLEGUIDE.md [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+---
+
+## <a id="Functions">Functions</a>
+
+### Prefer ternaries over simple `if/else` blocks
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:72]: But we should prefer `if/else` blocks over complex ternaries, because those are hard to read. 

#### Prefer
```ts
if (isBlah) {
 return 'a'
} else if (isFoo) {
  return 'b'
} else {
  return 'c'
}
```

over

```ts
return isBlah ? 'a' : isFoo ? 'b' : 'c'
```

File: STYLEGUIDE.md [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+if (isBlah) {
+  return a;
+} else {
+  return b;
+}
+```
+
+---
+
+### All new (exported) functions should have a JSDoc comment explaining functionality.
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:96]: Technically TSDoc

File: STYLEGUIDE.md [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+};
+```
+
+---
+
+### Prefer creating the function in the highest scope, and passing in all required data.
+
+#### Why
+
+Keeps functions pure.
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:139]: Can also consider creating utility closures (e.g. https://github.com/mongodb/leafygreen-ui/blob/943e4191dabb74acf403b2189327216085cc3e81/packages/brand-shape/src/utils.ts#L15)

File: STYLEGUIDE.md [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+---
+
+## <a id="Variable-naming">Variable Naming</a>
+
+### Acronyms are considered a single entity in camel-casing
+
+#### Why
+
+Some treat each letter in an acronym as its own entity when an acronyme is included in a var name. We'll instead consider the acronym as a single entity, and only capitalize the first letter of it (if that).
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:149]: This is opposite of most React variables/types though (e.g. `type React.HTMLElement`)

File: STYLEGUIDE.md [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+
+### Use UPPERCASE for static constants, and refactor to a shared file if needed.
+
+---
+
+### Use camelCase for most javascript functions and variables
+
+---
+
+### Use “be” verbs for boolean types. (e.g. should`/`can`/`does`/`has`)
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:173]: `isX`

File: STYLEGUIDE.md [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+let myWIPValue = 'pending';
+```
+
+---
+
+### Use UPPERCASE for static constants, and refactor to a shared file if needed.
+
+---
+
+### Use camelCase for most javascript functions and variables
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:169]: But use `PascalCase` for components/classes etc

File: STYLEGUIDE.md [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+
+### Avoid inline declaration of static variables when they can be moved to top of file or refactored into new file
+
+---
+
+### Prefer placing let/const variables at the beginning of their scope.
+
+---
+
+### Prefer const to let whenever possible.
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:187]: "`const` until you `can'tst`" 😆

File: STYLEGUIDE.md [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+---
+
+## <a id="Whitespace-and-line-breaks">Whitespace and line breaks</a>
+
+### Use a space before every prop defined in an interface
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:193]: Can probably lint this one

File: STYLEGUIDE.md [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+
+---
+
+## <a id="JSX-React">JSX/React</a>
+
+### Prefer factoring out any more-than-1-line event handlers into their own functions
+
+#### Why
+
+Minimizes inline javascript logic within
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:203]: Also creates a named function that'll show up in stack traces

File: STYLEGUIDE.md [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+return <div></div>;
+```
+
+---
+
+### Prefer using spread operator to avoid mutation of arrays and objects
+
+#### Why
+
+Immutability
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:273]: Not really sure what this means

File: STYLEGUIDE.md [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+
+---
+
+### Avoid nesting ternary operators
+
+---
+
+### Factor out of functional components any methods that do NOT require state
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:281]: Feels hard to parse. Would change to `"Factor out any methods that do NOT require state from functional components"`,  or even `"Prefer utility functions when full component state is not required"`

File: STYLEGUIDE.md (hunk 9 above)
Comment (TheSonOfThomp) [reply STYLEGUIDE.md:173]: Also, missing a tick before "should"

File: STYLEGUIDE.md (hunk 2 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:14]: Chatted offline, leaving this as-is for now. Can re-assess based on usage!

File: STYLEGUIDE.md (hunk 4 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:66]: Helps keeps files organized, and if we want to know where to look for things this will be helpful

File: STYLEGUIDE.md (hunk 5 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:72]: yes absolutely. I can document this as well.

Comment (brookescarlett) [reply STYLEGUIDE.md:72]: updated!

File: STYLEGUIDE.md (hunk 8 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:149]: Would you prefer WIP? I can't think of another example where we don't do that other than HTML

File: STYLEGUIDE.md (hunk 14 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:273]: Not mutating props that come in... creating a new object by spreading props.

File: STYLEGUIDE.md (hunk 12 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:193]: We don't currently

File: STYLEGUIDE.md (hunk 7 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:139]: What do you mean here?

**Comment by brookescarlett** [comment]
re-r? @TheSonOfThomp

File: STYLEGUIDE.md (hunk 8 above)
Comment (shaneeza) [reply STYLEGUIDE.md:149]: Hmm, I've never thought about this. It is a little easier to read when the entire acronym is capitalized. Are there examples of how other style guides handle this?

File: STYLEGUIDE.md (hunk 14 above)
Comment (shaneeza) [reply STYLEGUIDE.md:273]: Is there no example for this one?

File: STYLEGUIDE.md [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+
+#### Avoid
+
+```typescript
+<div className={css`...`} />
+```
+
+---
+
+### Postfix style variable names appropriately
```
Comment (shaneeza) [inline STYLEGUIDE.md:308]: Should we mention theme styles? `myComponentThemeStyle` and `myComponentBaseStyle`?

Comment (brookescarlett) [reply STYLEGUIDE.md:308]: Yeah! We should! What's the rule there? Maybe you should make a PR to my branch 😄

File: STYLEGUIDE.md (hunk 8 above)
Comment (brookescarlett) [reply STYLEGUIDE.md:149]: I took this from Atlas' style guide!

File: STYLEGUIDE.md (hunk 14 above)
Comment (TheSonOfThomp) [reply STYLEGUIDE.md:313]: Agh I see. I think an example on this one would help

File: STYLEGUIDE.md (hunk 7 above)
Comment (TheSonOfThomp) [reply STYLEGUIDE.md:171]: Sorry that was pretty unclear. Instead of defining a bunch of utilities that require component context, you can create a single utility function that accepts some component state as args, and returns a bunch of utilities with the component state in the closure. Not always useful, so it's fine to leave it out

File: STYLEGUIDE.md (hunk 8 above)
Comment (TheSonOfThomp) [reply STYLEGUIDE.md]: Most built-in examples use all caps (`CSS`, `XML`, `HTML`, `SVG`, `JSX`,  `JSON`, )
Though one example that's not even internally consistent is `XMLTHttpRequest`

Comment (brookescarlett) [reply STYLEGUIDE.md]: Kk, will leave this out

File: STYLEGUIDE.md [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+
+### Avoid nesting ternary operators
+
+---
+
+### Prefer utility functions when full component state is not required
+
+---
+
+## <a id="CSS-in-JS">CSS in JS</a>
```
Comment (TheSonOfThomp) [inline STYLEGUIDE.md:325]: @brookescarlett: @shaneeza mentioned documenting CSS property order. I'm personally a fan of the [Idiomatic CSS order](https://github.com/ream88/stylelint-config-idiomatic-order)

**Review by TheSonOfThomp — APPROVED** [review]
LGTM. Left a few comments

File: STYLEGUIDE.md (hunk 8 above)
Comment (TheSonOfThomp) [reply STYLEGUIDE.md]: I think that's fair. Not sure how much we use acronyms really (the only one that comes to mind is `FedRAMP`)
